File: python-algo/leetcode/problem_213.py
"""213. House Robber II
https://leetcode.com/problems/house-robber-ii/
"""
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        def dfs(i: int, flag: int) -> int:
            if memo[i][flag] != -1:
                return memo[i][flag]
            if i == len(nums) - 1:
                memo[i][flag] = nums[i] if flag == 0 else 0
            elif i == len(nums) - 2:
                memo[i][flag] = max(nums[i], (nums[i + 1] if flag == 0 else 0))
            else:
                memo[i][flag] = max(nums[i] + dfs(i + 2, flag),
                                    dfs(i + 1, flag))
            return memo[i][flag]

        n = len(nums)
        if n == 0:
            return 0
        if n == 1:
            return nums[0]
        memo = [[-1] * 2 for _ in range(len(nums))]
        return max(dfs(0, 1), dfs(1, 0))

    def rob2(self, nums: List[int]) -> int:
        n = len(nums)
        dp0 = [[0, 0] for _ in range(n)]
        for i in range(1, n):
            dp0[i][0] = max(dp0[i - 1])
            dp0[i][1] = dp0[i - 1][0] + nums[i]
        ans = max(dp0[-1])
        dp1 = [[0, 0] for _ in range(n)]
        dp1[0] = (0, nums[0])
        for i in range(1, n):
            dp1[i][0] = max(dp1[i - 1])
            dp1[i][1] = dp1[i - 1][0] + (0 if i == n - 1 else nums[i])
        return max(ans, max(dp1[-1]))

File: java-algo/src/main/java/com/leetcode/Problem213.java
package com.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * 213. House Robber II
 * https://leetcode.com/problems/house-robber-ii/
 *
 * You are a professional robber planning to rob houses along a street.
 * Each house has a certain amount of money stashed. All houses at this place
 * are arranged in a circle. That means the first house is the neighbor of the
 * last one. Meanwhile, adjacent houses have security system connected and it
 * will automatically contact the police if two adjacent houses were broken
 * into on the same night.
 *
 * Given a list of non-negative integers representing the amount of money of
 * each house, determine the maximum amount of money you can rob tonight
 * without alerting the police.
 *
 * Example 1:
 *
 * Input: [2,3,2]
 * Output: 3
 * Explanation: You cannot rob house 1 (money = 2) and then rob house 3 (money = 2),
 *              because they are adjacent houses.
 * Example 2:
 *
 * Input: [1,2,3,1]
 * Output: 4
 * Explanation: Rob house 1 (money = 1) and then rob house 3 (money = 3).
 *              Total amount you can rob = 1 + 3 = 4.
 */
public class Problem213 {
    public int rob(int[] nums) {
        int n = nums.length;
        if (n == 0) return 0;
        if (n == 1) return nums[0];
        if (n == 2) return Math.max(nums[0], nums[1]);
        // if rob the first house
        List<int[]> dp = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            dp.add(new int[2]);
        }
        dp.set(0, new int[]{nums[0], nums[0]});
        dp.set(1, new int[]{nums[0], nums[0]});
        for (int i = 2; i < n - 1; i++) {
            dp.get(i)[0] = Math.max(dp.get(i - 1)[0], dp.get(i - 1)[1]);
            dp.get(i)[1] = dp.get(i - 1)[0] + nums[i];
        }
        int ans = Math.max(dp.get(n - 2)[0], dp.get(n - 2)[1]);
        // if not rob the first house
        dp.set(0, new int[]{0, 0});
        for (int i = 1; i < n; i++) {
            dp.get(i)[0] = Math.max(dp.get(i - 1)[0], dp.get(i - 1)[1]);
            dp.get(i)[1] = dp.get(i - 1)[0] + nums[i];
        }
        return Math.max(ans, Math.max(dp.get(n - 1)[0], dp.get(n - 1)[1]));
    }

}
